### ISSUE TITLE
Centrage sur position

### ISSUE THREAD

**Issue by iulianpn** [PnX-SI/gn_mobile_occtax#241]
Bonjour à tous,

Nous sommes en cours des tests sur la version mobile 2.6.3210 et je voudrais savoir s'il était possible d'arriver sur l'écran Pointage directement sur le point géolocalisé et pas sur la zone de Clermont-Ferrand.

Est-il possible d'afficher directement cette zone (notre géolocalisation) par défaut ?

Merci et bonne après-midi

**Comment by sgrimault** [comment]
Bonjour @iulianpn,
Pour l'instant, on ne peut fixer la position par défaut que via le paramètre `map.center` (en lien aussi avec `map.max_bounds`). S'il n'est pas précisé, on pourrait effectivement utiliser la dernière position connue du terminal mais pour cela, ça implique que l'application "Occtax" possède déjà les permissions qui vont bien concernant la géolocalisation. Cette permission n'est demandée que si l'utilisateur souhaite être géolocalisé dans l'application (via le bouton de géolocalisation depuis l'étape de pointage)et que cette dernière n'a pas encore été octroyée.

**Comment by camillemonchicourt** [comment]
J'ai eu une autre demande récemment que la localisation GPS soit activée par défaut si on a déjà utilisé l'application.
A voir si c'est pertinent et/ou activable ou non dans la configuration à laquelle a accès l'utilisateur.

**Comment by iulianpn** [comment]
Bonsoir,
Merci pour votre réponse.
Plusieurs utilisateurs sur le terrain nous ont fait la remarque que si on pouvait passer sur l'écran Pointage avec une géolocalisation directe de leur position, ca peut leur faire gagner du temps non négligeable lors de la saisie (dans le cas ou il y'a affluence d'observations)
On pourrait peut-être ajouter une option dans le settings.json pour activer cette option.
Merci.

**Comment by iulianpn** [comment]
> J'ai eu une autre demande récemment que la localisation GPS soit activée par défaut si on a déjà utilisé l'application. A voir si c'est pertinent et/ou activable ou non dans la configuration à laquelle a accès l'utilisateur.

Effectivement encore mieux que chaque utilisateur puisse choisir son type de géolocalisation.

**Comment by camillemonchicourt** [comment]
Je ne suis pas certain de ce qui a été fait.
On se centre directement sur la position GPS de l'utilisateur seulement si le paramètre `map.center` n'est pas défini dans la config ?

Ce n'était pas vraiment la demande. Ce paramètre est quasiment tout le temps défini, voir tout le temps car il a une valeur par défaut il me semble.
L'objectif est que l'utilisateur puisse être directement centré sur sa position GPS pour lui faire gagner du temps dans sa saisie, comme aussi indiqué sur https://github.com/PnX-SI/gn_mobile_occtax/issues/245

**Comment by sgrimault** [comment]
Ah oui, je suis parti du fait que la résolution GPS se fait plutôt en "fallback" du paramètre `map.center` si ce dernier est non défini car ce paramètre reste optionnel.
On peut aussi le voir autrement où le paramètre `map.center` est en "fallback" de la résolution GPS (par exemple, l'utilisateur n'a pas accordé les permissions). Ce qui donnerait la résolution suivante :

```mermaid
---
config:
  title: Centrage automatique sur la carte
---
stateDiagram-v2
  direction TB

  state "position via le GPS" as gps
  state "permissions accordées ?" as perm
  state "position p" as position
  state "position c" as center
  state "via le paramètre 'map.center'" as param
  state "paramètre 'map.bounds' défini ?" as p_bounds
  state "paramètre 'map.bounds' défini ?" as c_bounds
  state "p ⊂ map.bounds ?" as p_in_bounds
  state "c ⊂ map.bounds ?" as c_in_bounds
  state "calcul du centroïde depuis map.bounds" as c_from_bounds
  state if_perm <<choice>>
  state if_param_center <<choice>>
  state if_param_p_bounds <<choice>>
  state if_param_c_bounds <<choice>>
  state if_p_in_bounds <<choice>>
  state if_c_in_bounds <<choice>>

  [*] --> gps
  gps --> perm
  perm --> if_perm
  if_perm --> position: oui et position résolue
  if_perm --> param: non
  position --> p_bounds
  p_bounds --> if_param_p_bounds
  if_param_p_bounds --> p_in_bounds: oui
  if_param_p_bounds --> [*]: p
  p_in_bounds --> if_p_in_bounds
  if_p_in_bounds --> [*]: p
  if_p_in_bounds --> param: non
  param --> if_param_center
  if_param_center --> center
  if_param_center --> [*]: non, position non défini
  center --> c_bounds
  c_bounds --> if_param_c_bounds
  if_param_c_bounds --> c_in_bounds: oui
  if_param_c_bounds --> [*]: c
  c_in_bounds --> if_c_in_bounds
  if_c_in_bounds --> [*]: c
  if_c_in_bounds --> c_from_bounds: non
  c_from_bounds --> [*]: c
```

**Comment by camillemonchicourt** [comment]
OK OK, je ne capte pas tout, mais ça me semble plus ça l'idée qui était demandée en effet.
Fait gagner du temps au maximum dans la saisie pour se centrer sur la position de l'utilisateur le plus vite possible, si possible.
Merci.

**Comment by sgrimault** [comment]
Oui, c'est le cas tel que je l'ai décrit ; GPS et sinon `map.center`. Donc dans le cas nominal si l'utilisateur a autorisé les permissions nécessaires, la carte sera automatiquement centrée sur la dernière position connue.

**Comment by gildeluermoz** [comment]
Je profite de ce ticket pour poser une question ,
Est-ce que dans le cas où les conditions suivantes sont réunies : 

- permissions accordées
- niveau de zoom minimal pour la saisie suffisant

on pointe le marqeur automatiquement sur la position GPS ?

Je ne suis pas sur de bien poser le sujet mais en gros je vois 2 modes de fonctionnement attendus par les utilisateurs : 

1. mode manuel : l'utilisateur souhaite naviguer dans la carte et placer le marqueur lui même indépendamment de la position GPS.
2. mode auto : l'utilisateur souhaite utiliser la position GPS pour positionner le marqueur.

Et du coup, je dirais que si le positionnement GPS est activé (avec les permissions accordée) et que la résolution est faite et peut-être aussi qu'elle est satisfaisante (`GPS_ACCURRACY` ?), on place le marqueur et on zoome sur la position en utilisant le `MIN_ZOOM_LEVEL`. Si l'utilisateur ne souhaite pas ce comportement, il désactive le GPS et passe en mode manuel.
Que ce soit sur ce prinicipe ou via un autre paramètre ou bouton explicite sur la carte, je pense qu'il serait pertinent de permettre un mode ou l'utilisateur considère que sa position, donc celle du GPS, correspond à celle de l'observation et que l'application lui économise l'étape de positionnement (zoomer) + pointage du marqueur.

Ce comportement "auto" devrait être conservé entre 2 relevés voir entre 2 sessions et activable/désactivable à l'étape "carte".

Qu'en pensez vous ?

**Comment by sgrimault** [comment]
Le plus simple serait simplement d'ajouter un autre bouton de saisie (au dessus du premier) qui permet d'ajouter directement le marqueur selon la position courante puis de recentrer la carte :
<img src="https://github.com/user-attachments/assets/fb83143c-a9d2-4381-807a-babee0bf83bb" width=256/>
Avec bien entendu la possibilité de le déplacer après coup.

**Comment by camillemonchicourt** [comment]
Non, à priori pas besoin, car cela va avec l'évolution à venir où le marqueur n'est plus à placer sur la carte mais au centre de la carte et c'est la carte qu'on bouge, comme cela se fait dans beaucoup d'outils et discuté ici : https://github.com/PnX-SI/gn_mobile_occtax/issues/138#issuecomment-2071715708

Donc on sera centré automatiquement sur sa position GPS et cela sera aussi la position initiale du marqueur, sans aucune autre action nécessaire par défaut.

**Comment by gildeluermoz** [comment]
Ok, parfait. Et le comportement de zommer sur une échelle qui permet de voir si le positionnement est bon, il reste manuel ?

**Comment by camillemonchicourt** [comment]
Par défaut, ça se centre et ça zoome sur sa position.
Il peut zoomer ou dézoomer manuellement. Il peut déplacer la carte et donc la localisation de l'observation en même temps au centre (et désactiver le centrage GPS).
Il peut valider, seulement si il au moins au niveau de zoom minimum configuré.

**Comment by camillemonchicourt** [comment]
Voici le fonctionnement en cours de mise en place pour la 2.8.0.
Avec centrage automatique, possibilité de placer au point au centre à tout moment avec la position GPS et/ou le centrage de la carte.
En conservant le fonctionnement actuel en complément, permettant de localiser le relevé avec un appui long sur la carte ou de déplacer le marqueur avec un appui long sur celui-ci : 

https://github.com/user-attachments/assets/b67a9038-afd0-474a-a512-33b4f0a66f75

**Comment by DonovanMaillard** [comment]
Fait dans la dernière version.
